### ISSUE TITLE
Unusual increase in the index of the inputBuffer in the AudioProcessor

### ISSUE THREAD

**Issue by nguyenhuynhdev** [androidx/media#925]
### Version

Media3 1.2.0

### Devices that reproduce the issue

* Emulator Pixel 7 Pro API 32, 33, 34.
* Infinix hot 30

### Devices that do not reproduce the issue

* Xiaomi K40s
* Samsung Note FE, S7Edge

### Reproducible in the demo app?

Not tested

### Reproduction steps

1. Utilize the queueInput function of BaseAudioProcessor or handleBuffer of TeeAudioProcessor, paying close attention to the index values of the inputBuffer.
2. Observe the unexpected and abnormal increase in the index values.

Here is my customized BaseAudioProcessor used to convert the inputBuffer into FFT:

```
    override fun onConfigure(inputAudioFormat: AudioProcessor.AudioFormat): AudioProcessor.AudioFormat {
        if (inputAudioFormat.encoding != C.ENCODING_PCM_16BIT) {
            throw AudioProcessor.UnhandledAudioFormatException(inputAudioFormat)
        }
        init(inputAudioFormat)
        return inputAudioFormat
    }

    override fun queueInput(inputBuffer: ByteBuffer) {
        val remaining = inputBuffer.remaining()
        if (remaining == 0) return
        processInputForFFT(inputBuffer)
        replaceOutputBuffer(remaining).put(inputBuffer).flip()
    }
    
    
    private fun init(inputAudioFormat: AudioProcessor.AudioFormat) {
        srcBufferPosition = 0
        bytesToProcess = SAMPLE_SIZE * 2
        dst = FloatArray(SAMPLE_SIZE + 2)
        sampleRate = inputAudioFormat.sampleRate
        channelCount = inputAudioFormat.channelCount
        audioTrackBufferSize = getDefaultBufferSizeInBytes(inputAudioFormat)
        srcBuffer = ByteBuffer.allocate(audioTrackBufferSize + BUFFER_EXTRA_SIZE)
        srcBuffer.clear()
    }
    
    private fun processInputForFFT(inputBuffer: ByteBuffer) {
        var position = inputBuffer.position()
        val limit = inputBuffer.limit()
        val fftBuffer = ByteBuffer.allocateDirect(inputBuffer.remaining()).order(ByteOrder.nativeOrder())
        while (position < limit) {
            var summedUp = 0
            for (channelIndex in 0 until channelCount) {
                summedUp += inputBuffer.getShort(position + 2 * channelIndex).toInt()
            }
            fftBuffer.putShort((summedUp / channelCount).toShort())
            position += channelCount * 2
        }
        srcBuffer.put(fftBuffer.array())
        srcBufferPosition += fftBuffer.array().size
        val tempShortArray = ShortArray(SAMPLE_SIZE)
        while (srcBufferPosition > audioTrackBufferSize) {
            srcBuffer.position(0)
            srcBuffer.asShortBuffer().get(tempShortArray, 0, SAMPLE_SIZE)
            val src =
                tempShortArray.map { (it.toFloat() * Short.MAX_VALUE + it) / (Short.MAX_VALUE * Short.MAX_VALUE) }
                    .toFloatArray()
            srcBuffer.position(bytesToProcess)
            srcBuffer.compact()
            srcBufferPosition -= bytesToProcess
            if (srcBufferPosition > 0) {
                srcBufferPosition = 0
            }
            srcBuffer.position(srcBufferPosition)
            // Call the function to process FFT (presumed to be a C++ function for Complex to FFT conversion)
            processFFT(src, dst)
        }
    }
    
    private fun getDefaultBufferSizeInBytes(inputAudioFormat: AudioProcessor.AudioFormat): Int {
        val outputPcmFrameSize =
            Util.getPcmFrameSize(inputAudioFormat.encoding, inputAudioFormat.channelCount)
        val minBufferSize =
            AudioTrack.getMinBufferSize(
                inputAudioFormat.sampleRate,
                Util.getAudioTrackChannelConfig(inputAudioFormat.channelCount),
                inputAudioFormat.encoding
            )
        Assertions.checkState(minBufferSize != AudioTrack.ERROR_BAD_VALUE)
        val multipliedBufferSize = minBufferSize * EXO_BUFFER_MULTIPLICATION_FACTOR
        val minAppBufferSize =
            durationUsToFrames(
                inputAudioFormat,
                EXO_MIN_BUFFER_DURATION_US
            ).toInt() * outputPcmFrameSize
        val maxAppBufferSize = minBufferSize.toLong()
            .coerceAtLeast(
                durationUsToFrames(
                    inputAudioFormat,
                    EXO_MAX_BUFFER_DURATION_US
                ) * outputPcmFrameSize
            )
            .toInt()
        val bufferSizeInFrames = Util.constrainValue(
            multipliedBufferSize,
            minAppBufferSize,
            maxAppBufferSize
        ) / outputPcmFrameSize
        return bufferSizeInFrames * outputPcmFrameSize
    }

    private fun durationUsToFrames(
        inputAudioFormat: AudioProcessor.AudioFormat,
        durationUs: Long
    ): Long {
        return durationUs * inputAudioFormat.sampleRate / C.MICROS_PER_SECOND
    }
    
    companion object {
        private const val SAMPLE_SIZE = 4096

        // From DefaultAudioSink.java:160 'MIN_BUFFER_DURATION_US'
        private const val EXO_MIN_BUFFER_DURATION_US: Long = 250000

        // From DefaultAudioSink.java:164 'MAX_BUFFER_DURATION_US'
        private const val EXO_MAX_BUFFER_DURATION_US: Long = 750000

        // From DefaultAudioSink.java:173 'BUFFER_MULTIPLICATION_FACTOR'
        private const val EXO_BUFFER_MULTIPLICATION_FACTOR = 4

        // Extra size next in addition to the AudioTrack buffer size
        private const val BUFFER_EXTRA_SIZE = SAMPLE_SIZE * 8
    }
```

### Expected result

https://github.com/androidx/media/assets/33225351/619f1687-f988-413f-9ee5-7759d55e83f1

### Actual result

[Screen_recording_20231228_162039.webm](https://github.com/androidx/media/assets/33225351/4994fe6c-f8c5-43ab-a69f-7f1689b6ba82)

### Media

*mp3

### Bug Report

- [ ] You will email the zip file produced by `adb bugreport` to android-media-github@google.com after filing this issue.

**Comment by nguyenhuynhdev** [comment]
das

**Comment by nguyenhuynhdev** [comment]
For those encountering similar issues when converting inputBuffer to FFT, there are a few scenarios to be aware of:
1. The entire FFT result may become NaN or infinity. (The error originates from the inputBuffer data)
2. The entire FFT result may spike to a random value, causing visual distortion. (The error is specific to certain devices when it comes to the inputBuffer data)
My temporary solution is to check the FFT result at position 2048. At this position, the value should oscillate within the range of 0 to 2. 
`jfloat _output = static_cast<float>(complex_abs(complexOutput[sampleSize / 2]));
if (std::isinf(_output) || std::isnan(_output) || _output > 2) {
    // Handle the problematic case
}`
This solution is temporary and impacts the performance of the application. Additionally, it may slow down the frame rate of the visualization.
I have thoroughly reviewed the entire code, and it seems that the issue is related to ExoPlayer. I am hopeful to receive a response from ExoPlayer developers soon. It has been almost two months since I reported this bug, and there has been no response.

**Comment by nguyenhuynhdev** [comment]
`@OptIn(UnstableApi::class)
class FFT(private val audioBufferSink: AudioBufferSink) {

    private var channelCount = 2
    private val dst = FloatArray(SAMPLE_SIZE/2)
    private val src = ShortArray(SAMPLE_SIZE )

    fun init(inputAudioFormat: AudioProcessor.AudioFormat) {
        channelCount = inputAudioFormat.channelCount
    }

    fun queueInput(inputBuffer: ByteBuffer) {
        var position = inputBuffer.position()
        val limit = inputBuffer.limit()
        src[0] = 0
        var i = 1
        while (position < limit) {
            src[i++] = inputBuffer.getShort(position + 2)
            position += channelCount * 2
        }
        processFFT(src, dst)
    }

    private external fun processFFT(input: ShortArray, output: FloatArray)

    fun onFftCallback(result: FloatArray) {
        audioBufferSink.onFft(result)
    }

    interface AudioBufferSink {
        fun onFft(fft: FloatArray)
    }

    companion object {
        init {
            System.loadLibrary("fft")
        }

        private const val TAG = "FFT"

        private const val SAMPLE_SIZE = 4096
    }
}`

[Screen_recording_20240214_012753.webm](https://github.com/androidx/media/assets/33225351/0b54ffe6-69d7-4fd0-8e33-7faa628ccb73)

**Comment by ychaparov** [comment]
Hi @nguyenhuynhdev !

Can you debug a bit on your end, and report back?
* For the same file, is the number of bytes given to you in `fun queueInput(inputBuffer: ByteBuffer)` different between the repro / no-repro devices?
* Are the byte values given to you in `fun queueInput(inputBuffer: ByteBuffer)` different between the repro / no-repro devices?

Thanks for sharing some code, but also seeing an example .mp3 file would help.

Pretty sure your code will only fill up `fftBuffer` completely for mono, 1-channel audio, while for 2-channel stereo files, `fftBuffer` will have a bunch of zeros when you convert it to array and put it in `srcBuffer`.

```
        val fftBuffer = ByteBuffer.allocateDirect(inputBuffer.remaining()).order(ByteOrder.nativeOrder())
        while (position < limit) {
            var summedUp = 0
            for (channelIndex in 0 until channelCount) {
                summedUp += inputBuffer.getShort(position + 2 * channelIndex).toInt()
            }
            fftBuffer.putShort((summedUp / channelCount).toShort())
            position += channelCount * 2
        }
        srcBuffer.put(fftBuffer.array())
```

**Comment by ychaparov** [comment]
When you say "the index values of the inputBuffer" -- what does that mean?
We give no guarantees about the size of `inputBuffer`s passed in to `AudioProcessor` -- the docs are here https://github.com/androidx/media/blob/b930b40a16c06318e43c81771fa2b1024bdb3f29/libraries/common/src/main/java/androidx/media3/common/audio/AudioProcessor.java#L163

The size of `inputBuffer`s can differ between devices with different audio decoder implementations

**Comment by nguyenhuynhdev** [comment]
Hello @ychaparov,

I appreciate your feedback. Upon investigating as per your suggestion, here are the refined results:

- The byte count for both repro and no-repro devices is consistent on the same audio file.
- Byte values differ, but only within the range of +-1.

I've identified this as an error originating from my FFT processing function. During debugging, I found that the C++ free() function, used to release complex data, is the cause of irregular fluctuations in the FFT on some devices. This is despite it being a data release function at the end without any logic code (still a problem in the Android native library). I've implemented an alternative solution for memory release, restoring stability. However, NaN or infinity errors persist during conversion, but these issues seem unrelated to the library, **so I'll close this bug report**.

Additionally, in the conversion process within the function:
```
var position = inputBuffer.position()
val limit = inputBuffer.limit()
src[0] = inputBuffer.getShort(position)
var i = 1
while (position < limit) {
    src[i++] = inputBuffer.getShort(position + 2)
    position += channelCount * 2
}
processFFT(src, dst)
```
To reduce the load on FFT processing and evenly distribute data from start to end, after obtaining the FFT, I will convert it once again to `frequencyBandLimits`. There is only a slight difference in the representation of the sound wave. For those interested in handling the details, they can use the following function to retrieve the entire audio data from both audio channels:

```
val array = inputBuffer.duplicate().moveToByteArray()
val srcBuff = ByteBuffer.allocateDirect(SAMPLE_SIZE * 2)
srcBuff.position(1)
srcBuff.put(array)
srcBuff.position(0)
srcBuff.asShortBuffer().get(src)
processFFT(src, dst)
```
The results are now really smooth:
[Screen_recording_20240217_234221.webm](https://github.com/androidx/media/assets/33225351/b1255f7a-ae7b-471f-ae9c-c800f834275b)

Thank you!
